File: LeetCode/20有效的括号.go
func isValid(s string) bool {
	hash := map[byte]byte{'(': ')', '[': ']', '{': '}'}
	var sta []byte
	if len(s)%2 != 0 {
		return false
	}
	for i := 0; i < len(s); i++ {
		if s[i] == '(' || s[i] == '[' || s[i] == '{' {
			sta = append(sta, hash[s[i]])
		} else if len(sta) > 0 && sta[len(sta)-1] == s[i] {
			sta = sta[:len(sta)-1]
		} else {
			return false
		}
	}
	return len(sta) == 0
}

File: LeetCode/20有效的括号.java
class Solution {
    public boolean isValid(String s) {
        if (s.length() % 2 != 0) return false;
        char[] chs = s.toCharArray();
        Stack<Character> sta = new Stack();
        for (int i = 0; i < chs.length; i++) {
            if (chs[i] == '(') sta.push(')');
            else if (chs[i] == '[') sta.push(']');
            else if (chs[i] == '{') sta.push('}');
            else if (sta.isEmpty() || sta.peek() != chs[i]) return false;
            else sta.pop();
        }
        return sta.isEmpty();
    }
}